src/mpid/ch4/generic/am/mpidig_am_msg.h

```diff
@@ -163,18 +179,27 @@ MPL_STATIC_INLINE_PREFIX int MPIDIG_recv_copy_seg(void *payload, MPI_Aint payloa
     MPIDIG_rreq_async_t *p = &(MPIDIG_REQUEST(rreq, req->async));
 
     if (p->recv_type == MPIDIG_RECV_DATATYPE) {
+        MPI_Aint max_data_size;
+        MPIR_Datatype_get_size_macro(MPIDIG_REQUEST(rreq, datatype), max_data_size);
+        max_data_size *= MPIDIG_REQUEST(rreq, count);
+
         MPI_Aint actual_unpack_bytes;
         MPIR_Typerep_unpack(payload, payload_sz,
                             MPIDIG_REQUEST(rreq, buffer),
                             MPIDIG_REQUEST(rreq, count),
                             MPIDIG_REQUEST(rreq, datatype), p->offset, &actual_unpack_bytes);
         p->in_data_sz -= actual_unpack_bytes;
         p->offset += actual_unpack_bytes;
-        if (payload_sz > actual_unpack_bytes) {
+        if (payload_sz + p->offset - actual_unpack_bytes > max_data_size) {
             /* did not fit */
             rreq->status.MPI_ERROR = MPIDIG_ERR_TRUNCATE(p->offset, p->offset + p->in_data_sz);
             MPIR_STATUS_SET_COUNT(rreq->status, p->offset);
             return 1;
+        } else if (payload_sz > actual_unpack_bytes) {
```

Comment:
Can we add a comment : ` /* basic element size mismatch */ `? I am fine with `basic type`, but I prefer `basic element size` because that is what the code checks, right?

Reply:
Fixed.